File: HoTT-Model/src/test/java/de/treichels/hott/model/enums/EnumTest.java
package de.treichels.hott.model.enums;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;

import java.io.File;
import java.io.IOException;
import java.lang.reflect.Field;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.net.URISyntaxException;
import java.net.URL;
import java.util.Enumeration;

import org.junit.Test;

public class EnumTest {
    @Test
    public void test() throws IOException, URISyntaxException, ClassNotFoundException, NoSuchMethodException, SecurityException, IllegalAccessException,
            IllegalArgumentException, InvocationTargetException {
        final Enumeration<URL> enums = Thread.currentThread().getContextClassLoader().getResources("gde/model/enums"); //$NON-NLS-1$

        while (enums.hasMoreElements()) {
            final File dir = new File(enums.nextElement().toURI());

            if (dir.isDirectory()) {
                final File[] files = dir.listFiles();

                for (final File file : files) {
                    final String fileName = file.getName();

                    if (fileName.endsWith(".class")) { //$NON-NLS-1$
                        final Class<?> clazz = Class.forName("gde.model.enums." + fileName.substring(0, fileName.length() - 6)); //$NON-NLS-1$

                        if (clazz.isEnum()) {
                            final Field[] fields = clazz.getDeclaredFields();
                            final Method toStringMethod = clazz.getMethod("toString"); //$NON-NLS-1$
                            final Method nameMethod = clazz.getMethod("name"); //$NON-NLS-1$

                            for (final Field field : fields) {
                                if ((field.getModifiers() & Modifier.STATIC & Modifier.PUBLIC) == 0) // skip non-public, non-static fields
                                    continue;

                                final Object object = field.get(null);
                                final String name = (String) nameMethod.invoke(object);
                                final String string = (String) toStringMethod.invoke(object);

                                assertEquals(field.getName(), name);
                                assertNotNull(string);
                            }
                        }
                    }
                }
            }
        }
    }
}

File: HoTT-Model/src/test/kotlin/de/treichels/hott/model/enums/EnumTest.kt
package de.treichels.hott.model.enums

import org.junit.Assert.*
import org.junit.Test
import java.io.File

class EnumTest {
    companion object {
        val packageNames = listOf("de.treichels.hott.model.enums", "de.treichels.hott.serial", "de.treichels.hott.model.voice")
    }

    @Test
    fun test() {
        packageNames.forEach { packageName ->
            Thread.currentThread().contextClassLoader
                    .getResources(packageName.replace(".", "/")).asSequence() // sequence of URLs
                    .map { File(it.toURI()) }.filter { it.isDirectory } // sequence of directories
                    .flatMap { it.listFiles().asSequence() }.filter { it.name.endsWith(".class") } // sequence of .class files
                    .map { "$packageName.${it.nameWithoutExtension}" }.map { Class.forName(it) }.filter { it.isEnum } // sequence of enums
                    .filter { it.declaredMethods.find { it.name == "toString" } != null } // with declared toString method
                    .forEach { clazz ->
                        val toStringMethod = clazz.getMethod("toString")
                        val nameMethod = clazz.getMethod("name")

                        clazz.declaredFields.filter { it.isEnumConstant }.forEach { field ->
                            val enumConstant = field.get(null)
                            val name = nameMethod.invoke(enumConstant) as String
                            val string = toStringMethod.invoke(enumConstant) as String

                            assertEquals(field.name, name)
                            assertNotNull(string)
                            assertNotEquals("[$name]", string)
                        }
                    }
        }
    }
}
